modules/core/src/main/java/org/apache/ignite/internal/client/thin/ReliableChannel.java

```diff
@@ -169,8 +167,10 @@
             // No-op.
         }
 
-        for (ClientChannelHolder hld : channels)
-            hld.closeChannel();
+        if (channels.get() != null) {
+            for (ClientChannelHolder hld: channels.get())
```

Comment:
Here and below: race condition - `channels` may become null after the null check. Copy to a variable, then check for null / iterate / etc.

Reply:
There is no race conditions. As channels are set within synchronized method `initChannelHolders`. The discussed code is placed within `close` that is synchronized too. Also channels is never reference to null after it initialized, so this check is enough.

For other accessors to the `channels` field:
1. `rollCurrentChannel` invokes after channels is initialized. Also it guards with `curChannelsGuard.writeLock`, so it is safe the reference directly;
2. `applyOnDefaultChannel` can lead to NPE in case `initConnection` is not invoked. So I've just lazy initialization in the method, so there is a guarantee that channels are initialized. In our case this method call just returns fast;
3. `initAllChannelsAsync` are invoked after channels are initialized. Also it uses `ArrayList.iterator()` to iterate through channels. Under the hood it just creates Iterator object that. We change `channels` by reference switch so, this iterator is safe to use.
4. `channelsInit` is checks channels for null and then initialize them is it needed. So it's safe too.

Reply:
Sure, with current usages it may work, but the approach is fragile and hard to trace. See how you had to check every usage to make sure this logic is safe?

For volatiles and atomics it is a common practice to copy the value into a variable when you do multiple access and/or conditional logic:
* Easy to see that it is safe - less cognitive load
* Faster

Reply:
Fixed. The only place channels used directly is a single getter in `applyOnDefaultChannel` within guard read lock.